accord-core/src/main/java/accord/local/CommandStores.java

```diff
@@ -474,40 +474,42 @@ public  <O> void mapReduceConsume(PreLoadContext context, IntStream commandStore
 
     public <O> AsyncChain<O> mapReduce(PreLoadContext context, Routables<?> keys, long minEpoch, long maxEpoch, MapReduce<? super SafeCommandStore, O> mapReduce)
     {
-        AsyncChain<O> chain = null;
         BiFunction<O, O, O> reducer = mapReduce::reduce;
         Snapshot snapshot = current;
         ShardHolder[] shards = snapshot.shards;
+        List<AsyncChain<? extends O>> chains = null;
```

Comment:
why?

Reply:
feedback from Ariel. There is a bug with combiner where we double called the callback, because the callback threw. I added a TODO documenting this bug in the original feedback PR and he asked me to just go and fix it...

All / Reduce / ReduceWithIdentiy were all acting like .map logic, so rather than making everything handle exceptions properly it was simpler to centralize exception handling and replace all 3 cases with a .map call; this makes Reduce class no longer exist.

So for these call paths, had to move the List<AsyncChain<? extends O>> chains from Reduce to the caller.

Reply:
Wouldn't a simpler solution have been

```
    abstract O complete(I[] results);
```

and
 
```
        if (REMAINING.decrementAndGet(this) == 0)
        {
            O combinedResult;
            try
            {
                combinedResult = complete(results());
            }
            catch (Throwable t)
            {
                callback.accept(null, t);
                return;
            }
            callback.accept(combinedResult, null);
        }
```
?

This also maintains the benefit of not encapsulating when there's only one async operation - which we expect to be the norm.

Reply:
>   abstract O complete(I[] results);

I thought about that, but went with not doing that as that is reinventing `map`... so can delete all 3 classes in favor of reusing the existing `.map` logic (and centralize the exception handling).

> This also maintains the benefit of not encapsulating when there's only one async operation - which we expect to be the norm.

Yes, in the case where there is a single chain this is slightly more memory than before.  In this case the memory is short lived and stack local (so should be relatively cheap).

Reply:
It is explicitly not stack local?

Reply:
ill send a patch to add the same semantic, while keeping the genetical feedback in-tact.

Reply:
pushed commit adding the old semantic back for the cases that called `reduce(chain1, chain2, reducer)`

Reply:
I’d be fine with a version of the loop that just allocates the list on encountering the second chain